File: src/services/remoteaccessservice.cpp
#include "services/remoteaccessservice.h"

#include <QDir>
#include <QUrl>

RemoteAccessService::RemoteAccessService(QObject *parent)
    : QObject(parent)
{
}

bool RemoteAccessService::busy() const
{
    return m_process != nullptr;
}

QString RemoteAccessService::buildUri(const QString &protocol, const QString &host,
                                      const QString &remotePath, const QString &user,
                                      int port, const QString &share) const
{
    if (host.trimmed().isEmpty())
        return {};

    QUrl url;
    url.setScheme(protocol.trimmed().toLower());
    url.setHost(host.trimmed());
    if (!user.trimmed().isEmpty())
        url.setUserName(user.trimmed());
    if (port > 0)
        url.setPort(port);

    QString path = remotePath.trimmed();
    if (url.scheme() == QStringLiteral("smb")) {
        const QString trimmedShare = share.trimmed();
        if (!trimmedShare.isEmpty())
            path = trimmedShare + (path.isEmpty() ? QString() : QStringLiteral("/") + path);
    }

    if (path.isEmpty())
        path = QStringLiteral("/");
    if (!path.startsWith('/'))
        path.prepend('/');

    url.setPath(QDir::cleanPath(path));
    return url.toString(QUrl::FullyEncoded);
}

void RemoteAccessService::connectToLocation(const QString &uri)
{
    const QString trimmedUri = uri.trimmed();
    if (trimmedUri.isEmpty()) {
        emit connectionFinished(false, QString(), QStringLiteral("Enter a remote location"));
        return;
    }

    if (m_process) {
        // Its signals must not reach the handlers below once a new process
        // owns m_process; ~QProcess emits them synchronously on kill.
        disconnect(m_process, nullptr, this, nullptr);
        m_process->kill();
        m_process->deleteLater();
        m_process = nullptr;
    }

    m_pendingUri = trimmedUri;
    QProcess *process = new QProcess(this);
    m_process = process;
    emit busyChanged();

    // errorOccurred(Crashed) and finished() fire for the same event; whichever
    // runs first settles the connection and the other bails on the guard.
    auto settle = [this, process](bool success, const QString &error) {
        if (process != m_process)
            return;
        emit connectionFinished(success, m_pendingUri, error);
        process->deleteLater();
        m_process = nullptr;
        m_pendingUri.clear();
        emit busyChanged();
    };

    connect(process, qOverload<int, QProcess::ExitStatus>(&QProcess::finished), this,
            [process, settle](int exitCode, QProcess::ExitStatus) {
        settle(exitCode == 0, exitCode == 0
            ? QString()
            : QString::fromUtf8(process->readAllStandardError()).trimmed());
    });

    connect(process, &QProcess::errorOccurred, this, [process, settle](QProcess::ProcessError) {
        settle(false, process->errorString());
    });

    process->start(QStringLiteral("gio"), {QStringLiteral("mount"), trimmedUri});
}

File: crates/bubble-app/src/services/remote_access_service.rs
use qmetaobject::*;
use std::process::Command;

#[allow(non_snake_case)]
#[derive(QObject, Default)]
pub struct RemoteAccessService {
    _base: qt_base_class!(trait QObject),

    pub busy: qt_property!(bool; NOTIFY busyChanged),
    pub busyChanged: qt_signal!(),
    pub connectionFinished: qt_signal!(success: bool, uri: QString, error: QString),

    pub buildUri: qt_method!(fn(&self, protocol: QString, host: QString, remote_path: QString, user: QString, port: i32, share: QString) -> QString),
    pub connectToLocation: qt_method!(fn(&mut self, uri: QString)),
}

impl RemoteAccessService {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn buildUri(
        &self,
        protocol: QString,
        host: QString,
        remote_path: QString,
        user: QString,
        port: i32,
        share: QString,
    ) -> QString {
        let host_str = host.to_string().trim().to_string();
        if host_str.is_empty() {
            return QString::default();
        }
        let proto_str = protocol.to_string().trim().to_lowercase();
        let user_str = user.to_string().trim().to_string();
        let share_str = share.to_string().trim().to_string();
        let path_str = remote_path.to_string().trim().to_string();

        let full_path = if proto_str == "smb" {
            if !share_str.is_empty() {
                if path_str.is_empty() {
                    format!("/{}", share_str)
                } else if path_str.starts_with('/') {
                    format!("/{}{}", share_str, path_str)
                } else {
                    format!("/{}/{}", share_str, path_str)
                }
            } else if path_str.starts_with('/') {
                path_str
            } else if path_str.is_empty() {
                "/".to_string()
            } else {
                format!("/{}", path_str)
            }
        } else if path_str.starts_with('/') {
            path_str
        } else if path_str.is_empty() {
            "/".to_string()
        } else {
            format!("/{}", path_str)
        };

        let mut uri = format!("{}://", proto_str);
        if !user_str.is_empty() {
            uri.push_str(&user_str);
            uri.push('@');
        }
        uri.push_str(&host_str);
        if port > 0 {
            uri.push_str(&format!(":{}", port));
        }
        uri.push_str(&full_path);
        QString::from(uri.as_str())
    }

    pub fn connectToLocation(&mut self, uri: QString) {
        let uri_str = uri.to_string().trim().to_string();
        if uri_str.is_empty() {
            self.connectionFinished(false, QString::default(), QString::from("Enter a remote location"));
            return;
        }

        self.busy = true;
        self.busyChanged();

        let output = Command::new("gio")
            .arg("mount")
            .arg(&uri_str)
            .output();

        self.busy = false;
        self.busyChanged();

        match output {
            Ok(out) => {
                let success = out.status.success();
                let err_str = String::from_utf8_lossy(&out.stderr).trim().to_string();
                self.connectionFinished(success, QString::from(uri_str.as_str()), QString::from(err_str.as_str()));
            }
            Err(e) => {
                self.connectionFinished(false, QString::from(uri_str.as_str()), QString::from(e.to_string().as_str()));
            }
        }
    }
}
